### PR TITLE
Add kubernetes example for hotrod app

### PR DESCRIPTION
**Pull request by highb**
## Which problem is this PR solving?
- Provides an easy example for running the Jaeger all-in-one service and the `hotrod` app in any k8s cluster for demo and testing purposes.

## Short description of the changes
- Adds a kubernetes deployment example with configuration modes for Jaeger all-in-one and the `hotrod` app by itself. Can be used with `kustomize` out of the box or the manifests can be applied directly to your cluster.

### CODE REVIEW

File: examples/hotrod/README.md [hunk 1]
```diff
@@ -65,6 +65,15 @@ docker run \
 
 Then open http://127.0.0.1:8080
 
+### Run HotROD from kubernetes
+
+```bash
+kustomize build ../kubernetes | kubectl apply -f -
```
Comment (yurishkuro) [inline examples/hotrod/README.md:71]: examples/ folder is not limited to hotrod, couldn't this be under examples/hotrod/kubernetes ?

File: examples/kubernetes/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: example-hotrod
+  name: example-hotrod
+  namespace: example-hotrod
+spec:
```
Comment (yurishkuro) [inline examples/kubernetes/deployment.yaml:8]: I don't see this running the actual Jaeger backend.

File: examples/kubernetes/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+  - namespace.yaml
+  - deployment.yaml
+  - service.yaml
+
+# If you want to use the grafana-agent, uncomment below.
```
Comment (yurishkuro) [inline examples/kubernetes/kustomization.yaml:9]: not relevant

**Review by yurishkuro — CHANGES_REQUESTED** [review]

File: examples/kubernetes/kustomization.yaml (hunk 3 above)
Comment (highb) [reply examples/kubernetes/kustomization.yaml:9]: Fair enough, I'll remove it.

File: examples/kubernetes/deployment.yaml (hunk 2 above)
Comment (highb) [reply examples/kubernetes/deployment.yaml:8]: The use-case I have in mind for this is that you already have some tracing backend working and you need to test it out with an example app. I could make a different config that also deploys the all-in-one Jaeger alongside the app.

Comment (joe-elliott) [reply examples/kubernetes/deployment.yaml:8]: I love the idea of a k8s example, but it really needs to start up the jaeger as well. That way people just getting into Jaeger can easily spin up an example app and functioning backend and see the basic configuration points to get things working.

File: examples/kubernetes/kustomization.yaml (hunk 3 above)
Comment (highb) [reply examples/kubernetes/kustomization.yaml]: Removed in https://github.com/jaegertracing/jaeger/pull/3645/commits/d09add6e1d564c03e471d2f0d5a1841684cc72a7

File: examples/kubernetes/deployment.yaml (hunk 2 above)
Comment (highb) [reply examples/kubernetes/deployment.yaml]: I agree. Added in https://github.com/jaegertracing/jaeger/pull/3645/commits/12122dd398c40fa59f9fd23e0710732c1b61c4c3 however I'm not familiar with how the all-in-one is supposed to work. I have it running but the traces don't seem to be showing up. If I have some time later I'll poke around with it a bit more.

File: examples/hotrod/README.md (hunk 1 above)
Comment (highb) [reply examples/hotrod/README.md]: Moved in https://github.com/jaegertracing/jaeger/pull/3645/commits/4e7ce011354c2dbbfd6eb28f077d32e92423d88c

File: examples/kubernetes/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Hot R.O.D. - Rides on Demand on Kubernetes
+
+Example k8s manifests for deploying the [hotrod app](../hotrod) to your k8s environment of choice. e.g. minikube, k3s, EKS, GKE
+
+## Features
+
+- Optional [example configuration](kustomize.yaml) for using [grafana-agent-traces](https://grafana.com/docs/grafana-cloud/traces/set-up-and-use-tempo/) as a backend.
```
[the remark this answers is not in the document]
Comment (highb) [reply examples/kubernetes/README.md]: Removed in https://github.com/jaegertracing/jaeger/commit/d09add6e1d564c03e471d2f0d5a1841684cc72a7

File: examples/kubernetes/deployment.yaml (hunk 2 above)
Comment (highb) [reply examples/kubernetes/deployment.yaml]: Oh! I think I got it. The port needs to be `UDP` not `TCP`. Testing that out now.

Comment (highb) [reply examples/kubernetes/deployment.yaml]: Nope, that doesn't seem to be it. Will try again later.

File: examples/hotrod/README.md [hunk 5]
```diff
@@ -65,6 +65,17 @@ docker run \
 
 Then open http://127.0.0.1:8080
 
+### Run HotROD/Jaeger all-in-one from kubernetes
+
+```bash
+kustomize build kubernetes | kubectl apply -f -
```
Comment (yurishkuro) [inline examples/hotrod/README.md:71]: nit: make it more obvious that it's referring to a directory
```suggestion
kustomize build ./kubernetes | kubectl apply -f -
```

File: examples/hotrod/README.md [hunk 6]
```diff
@@ -65,6 +65,17 @@ docker run \
 
 Then open http://127.0.0.1:8080
 
+### Run HotROD/Jaeger all-in-one from kubernetes
```
Comment (yurishkuro) [inline examples/hotrod/README.md:68]: I would move this right after the section `### Run everything via docker-compose`

File: examples/hotrod/kubernetes/base/hotrod/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+          - containerPort: 8083
+            name: route
+        resources:
+          limits:
+            cpu: 100m
+            memory: 100M
+          requests:
+            cpu: 100m
+            memory: 100M
+      - image: redis:latest
```
Comment (yurishkuro) [inline examples/hotrod/kubernetes/base/hotrod/deployment.yaml:41]: why is this needed?

Comment (highb) [reply examples/hotrod/kubernetes/base/hotrod/deployment.yaml:41]: The `hotrod` app appears to need `redis`. When I was viewing the traces it was failing to connect.

File: examples/hotrod/README.md (hunk 6 above)
Comment (highb) [reply examples/hotrod/README.md]: Done in https://github.com/jaegertracing/jaeger/pull/3645/commits/05330372471e57e49436af21d738741bd77674f1

File: examples/hotrod/README.md [hunk 8]
```diff
@@ -28,6 +28,17 @@ to view the traces. A tutorial / walkthrough is available:
 
 Alternatively, you can run each component separately as described below.
 
+### Run HotROD/Jaeger all-in-one from kubernetes
```
Comment (yurishkuro) [inline examples/hotrod/README.md:31]:
```suggestion
### Run everything in Kubernetes
```

File: examples/hotrod/README.md [hunk 9]
```diff
@@ -28,6 +28,17 @@ to view the traces. A tutorial / walkthrough is available:
...
+
+```bash
+kustomize build ./kubernetes | kubectl apply -f -
+kubectl port-forward -n example-hotrod service/example-hotrod 8080:frontend
+# In another terminal
+kubectl port-forward -n example-hotrod service/jaeger 16686:frontend
+```
+
+Then open <http://127.0.0.1:8080>
```
Comment (yurishkuro) [inline examples/hotrod/README.md:40]:
```suggestion
Access Jaeger UI at http://localhost:16686 and HotROD app at http://localhost:8080
```

File: examples/hotrod/README.md [hunk 10]
```diff
@@ -28,6 +28,17 @@ to view the traces. A tutorial / walkthrough is available:
...
+
```
Comment (yurishkuro) [inline examples/hotrod/README.md:41]: how to clean up?

File: examples/hotrod/kubernetes/README.md [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Hot R.O.D. - Rides on Demand on Kubernetes
+
+Example k8s manifests for deploying the [hotrod app](..) to your k8s environment of choice. e.g. minikube, k3s, EKS, GKE
+
+## Deploy with Kustomize
+
+```
+kustomize build . | kubectl apply -f -
```
Comment (yurishkuro) [inline examples/hotrod/kubernetes/README.md:8]: there are more steps in the top README

File: examples/hotrod/kubernetes/base/hotrod/deployment.yaml (hunk 7 above)
Comment (yurishkuro) [reply examples/hotrod/kubernetes/base/hotrod/deployment.yaml:41]: Hotrod binary has no external dependencies, it fakes them.

Comment (highb) [reply examples/hotrod/kubernetes/base/hotrod/deployment.yaml:41]: Gotcha. I'll remove redis, then.

File: examples/hotrod/kubernetes/README.md (hunk 11 above)
Comment (highb) [reply examples/hotrod/kubernetes/README.md:8]: Copied over steps from top README with modified paths.

File: examples/hotrod/README.md (hunk 10 above)
Comment (highb) [reply examples/hotrod/README.md:44]: Added https://github.com/jaegertracing/jaeger/pull/3645/commits/5477e7e5cfeabe581042100ef0c230bbebc0248a

File: examples/kubernetes/deployment.yaml (hunk 2 above)
Comment (highb) [reply examples/kubernetes/deployment.yaml]: Got it working by moving the jaeger all-in-one container in to the same pod and using localhost as the jaeger URL. 🎉

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
Thanks, @highb. If people find issues with these instructions, prepare to be tagged :-)
